### PR TITLE
Openstack node template enhancements

### PR DESCRIPTION
**Pull request by thpham**
Hello,

This pull request is an attempt to enhance two features requested in the following issues:

1. [Openstack Node Driver BootFromVolume](https://github.com/rancher/rancher/issues/27270)
2. [Openstack Node Template - Block device mapping](https://github.com/rancher/rancher/issues/24319)

---

1. When `openstack-boot-from-volume` is true, instead of using ephemeral volume, it requests the creation of a cinder volume of type `openstack-volume-type` and size `openstack-volume-size` for the ROOT system volume.

2. When `openstack-boot-from-volume` is false and `openstack-volume-size` is greater than 0, then it requests the creation of an additional cinder volume of type `openstack-volume-type` that will be attached as another device (/dev/vdb). In this situation the root volume remains on ephemeral volume. This allows to mount dedicated storage (from cloud-init) and use that host storage for rancher longhorn (example).

I was not able to fully test it on my environment yet. But I would love to get feedback first. If someone is able to test, I can provide the corresponding rancher image build with these features.

Thank for reviewing !

Cheers

### CODE REVIEW

**Comment by qkrtjddlsgh** [comment]
Is there a way to use Openstack Block device mapping and BootFromVolume in Rancher 2.X? not docker/machine.

I'm using rancher/rancher:latest docker image.

Thank you.

**Comment by thpham** [comment]
@qkrtjddlsgh @AliBeddah This PR is exactly what I try to include in rancher 2.x, rancher include rancher-machine for Openstack node-template. I built [a specific image of rancher ](https://hub.docker.com/repository/docker/tpham/rancher/tags?page=1) that include the changes of this PR for testing: [here is the commit](https://github.com/thpham/rancher/commit/5da0ac1525543389b1aea5c1f0ff1d75d8412222).

You can use the image: `tpham/rancher:5da0ac152`  if you want to test and give feedback.

Then you can create a node-template with the option and following what I described previously (point 1 / 2).
Here is a screenshot: https://imgur.com/WfgnZAE.png

Thanks helping me testing that !

**Comment by qkrtjddlsgh** [comment]
Thanks for your answer. I have one more question.

Can I add a --port or --nic field to the openstack config?
I will use openstack instance by sriov.

**Comment by AliBeddah** [comment]
Thank you @thpham for your contribution. VMs booted from volume, but I got this error while deploying cluster nodes.

Unable to find image 'rancher/rancher-agent:5da0ac152' locally docker: Error response from daemon: manifest for rancher/rancher-agent:5da0ac152 not found: manifest unknown: manifest unknown. See 'docker run --help'. Process exited with status 125 : exit status 1
--

image

**Comment by thpham** [comment]
@AliBeddah thanks for testing.

With the previous rancher image I just retagged the produced image,  and I forgot to build the image from my branch with the env variable `REPO`. 

now I did:

```
export REPO=tpham
make
```

and it produced 2 images which I pushed on docker hub.

```
tpham/rancher:dev
tpham/rancher-agent:dev
```

Could you tried again with `tpham/rancher:dev` image which could pull the agent when you create a cluster.

Keep me posted, thanks.

**Comment by Pure-AdamuKaapan** [comment]
@thpham any updates on this? I can give it a quick test if you like, I just stumbled into a use case where I need this again. Is `tpham/rancher:dev` still a build that contains this fix?

**Comment by Pure-AdamuKaapan** [comment]
@thpham I've attempted to test this, but it's currently failing with this:
`error syncing 'c-8kgmt/m-4mkzc': handler node-controller: Error creating machine: Error in driver during machine creation: No suitable endpoint could be found in the service catalog., requeuing`

After some digging, I believe it's because you're using `openstack.NewBlockStorageV1` instead of `V2` or `V3`. I'll make a test build with V3 and see what happens.

In the meantime, since I'm not super familiar with Rancher code, what's the usual procedure for version checking?

**Comment by Pure-AdamuKaapan** [comment]
Confirmed, updating that to BlockStorageV3 fixes the issue I have. I will note that the volume type parameter no longer worked, probably due to this upgrade. I may poke around a bit more to see if it fixes things.

File: drivers/openstack/client.go
Comment (Pure-AdamuKaapan) [inline drivers/openstack/client.go:109]:
```suggestion

```

File: drivers/openstack/client.go
Comment (Pure-AdamuKaapan) [inline drivers/openstack/client.go:572]:
```suggestion
	blockStorage, err := openstack.NewBlockStorageV3(c.Provider, gophercloud.EndpointOpts{
```

File: drivers/openstack/client.go
Comment (Pure-AdamuKaapan) [reply drivers/openstack/client.go:109]: In the gophercloud docs, `This requires Compute API microversion 2.67 or later`. Tested on an OpenStack setup with compute min microversion 2.1 and max microversion 2.79 (and for what it's worth, block min 3.0 and max 3.59).

**Review by Pure-AdamuKaapan — COMMENTED** [review]
Left comments with the changes I had to make to get it to work on my setup. YMMV I guess, but figured I'd leave them here in case it helped

**Comment by janeczku** [comment]
@Pure-AdamuKaapan Thanks for testing the PR. Would be great to land this feature in the code. Waiting on feedback from @thpham whether he is going to update this PR so we have a working version to merge.

File: drivers/openstack/client.go
Comment (thpham) [reply drivers/openstack/client.go:109]: Hello, this is something I would really need, I've a use case where ROOT volume need to be on a "ssd" disk pool.
(ex. a node-template for k8s master where etcd requires fast I/O.)
As you mention this requires specific `Compute API microversion 2.67 or later`, but if you leave it empty it doesn't break and choose you default volume_type/pool.

**Comment by thpham** [comment]
Thank you @Pure-AdamuKaapan for helping me to review this PR.
I was able to test it over this weekend and it looks like the both use case of this PR works.

To sumarize:

1. Openstack Node Driver BootFromVolume:

```
rancher-machine --debug create -d openstack  \
--openstack-boot-from-volume=true \
--openstack-volume-size=20 \
test-machine
```

This only works with `Compute API microversion 2.67 or later`

```
rancher-machine --debug create -d openstack  \
--openstack-boot-from-volume=true \
--openstack-volume-type=ssd \
--openstack-volume-size=20 \
test-machine
```

2. Openstack Node Template - Block device mapping:

it creates an additional cinder volume (type=ssd) attached to the instance.

```
rancher-machine --debug create -d openstack  \
--openstack-volume-type=ssd \
--openstack-volume-size=20 \
test-machine
```

**Comment by janeczku** [comment]
@thpham What is the default volume type if not specified or pre Compute API microversion 2.67 ?

**Comment by thpham** [comment]
@janeczku  the `default_volume_type` is proper to each openstack installation and the configuration you have in `cinder.conf`
cf. https://docs.openstack.org/cinder/latest/cli/cli-manage-volumes.html

File: drivers/openstack/client.go
Comment (Pure-AdamuKaapan) [reply drivers/openstack/client.go:109]: Ahhh, got it, I think I misread some code and thought this wasn't necessary for booting. Still learning, thanks for the clarification!

What's the best way to force my rancher-machine to use this version of the compute API? My setup seems to support it but I guess it's not using that API version for some reason?

Comment (thpham) [reply drivers/openstack/client.go:109]: Do you have access to an Openstack Stein version or later ?

File: drivers/openstack/client.go
Comment (thpham) [inline drivers/openstack/client.go:117]:
```suggestion
		}
		if d.VolumeType != "" {
			c.Compute.Microversion = "2.67"
		}
		server, err = bootfromvolume.Create(c.Compute, serverOpts).Extract()
```

Comment (janeczku) [reply drivers/openstack/client.go:117]: Seems to be the correct approach to force the server to return the required API version 👍

Comment (Pure-AdamuKaapan) [reply drivers/openstack/client.go:117]: I like it, makes sense to me.

File: drivers/openstack/client.go
Comment (Pure-AdamuKaapan) [reply drivers/openstack/client.go:109]: I have access to a Train cluster I believe. If you make a new test build with your latest compute version change I can give it a try pretty quick

Comment (thpham) [reply drivers/openstack/client.go:109]: I think it's ok now. I tested it against a lower version and I saw the API rejecting the instance creation and I saw specifically the client had the 2.67 microversion header.

**Comment by thpham** [comment]
LGTM now, ready for final approval ! thank you all.

File: drivers/openstack/client.go
Comment (Pure-AdamuKaapan) [reply drivers/openstack/client.go:109]: @thpham do you have a new build at your `thpham/rancher:dev` tag?

File: drivers/openstack/openstack.go
Comment (Pure-AdamuKaapan) [inline drivers/openstack/openstack.go:526]: Just to validate, this is for the "create a VM with a block volume, but not booting from it" case?

**Review by Pure-AdamuKaapan — APPROVED** [review]
LGTM! Note I'm not a rancher-machine developer, just a stakeholder in the PR.

File: drivers/openstack/openstack.go
Comment (thpham) [reply drivers/openstack/openstack.go:526]: yes, correct !

File: drivers/openstack/client.go
Comment (thpham) [reply drivers/openstack/client.go:109]: here it is, the new image (based on rancher:master) with the current changes of this PR: `tpham/rancher:8acfe59c2`

Comment (Pure-AdamuKaapan) [reply drivers/openstack/client.go:109]: Sweet, giving it a go now.

Comment (Pure-AdamuKaapan) [reply drivers/openstack/client.go:109]: Worked for me! Volume type was set correctly, no errors thrown, VM provisioned properly.

**Comment by Pure-AdamuKaapan** [comment]
Bump on this, I assume we're just waiting for review from @luthermonson ?

**Comment by janeczku** [comment]
Working on it @Pure-AdamuKaapan

**Comment by thpham** [comment]
Is there anything we can do to ease the review ? I'm also doing this  evaluation for the company I'm working with, we are potentialy also interested to go forward with Rancher product.
But, is there any different priorities given if you are a paying customer or simply a free contributor ?

**Comment by xinzhongyoumeng** [comment]
LGTM!

**Comment by xinzhongyoumeng** [comment]
> Working on it @Pure-AdamuKaapan

Is there any progress？

**Review by dramich — APPROVED** [review]

**Comment by luthermonson** [comment]
@thpham and @janeczku thanks for your work on this, great PR

**Comment by realmcarlin** [comment]
Hello guys thank you for your work, I am working for a company that want to use ranger and this feature is mandatory for us.

**Pull request merged**
